flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala

```diff
@@ -411,6 +411,23 @@ class TaskManager(
             log.debug(s"Cannot find task to fail for execution ${executionID})")
           }
 
+        // stops a task
+        case StopTask(executionID) =>
+          val task = runningTasks.get(executionID)
+          if (task != null) {
+            try {
+              task.stopExecution()
```

Comment:
Well. We know, that stop is only sent to sources, thus, the user function is of type `SourceFunction`. The `cancel()` method has the following JavaDoc:
```
	/**
	 * Cancels the source. Most sources will have a while loop inside the
	 * {@link #run} method. You need to ensure that the source will break out of this loop. This
	 * can be achieved by having a volatile field "isRunning" that is checked in the loop and that
	 * is set to false in this method.
	 */
	void cancel();
```
I interpreted this a a non-blocking function. Even if this requirement is not specified explicitly. We could just do this (need to talk to other streaming folk about implication though). Or we add a new method `stop()` to `SourceFunction`. However, this implies to change a lot of code... All classes implementing `SourceFunction` need to get extended (this was another reason why I just used `cancel()`).